core/arch/arm/plat-rockchip/psci_rk322x.c

```diff
@@ -43,6 +43,175 @@
 #include <tee/entry_std.h>
 #include <tee/entry_fast.h>
 
+struct dram_data {
+	uint32_t cru_mode_con;
+	uint32_t cru_clksel0;
+	uint32_t cru_clksel1;
+	uint32_t cru_clksel10;
+	uint32_t cru_clksel21;
+	uint32_t cru_clkgate[CRU_CLKGATE_CON_CNT];
+};
+
+static struct dram_data dram_d;
+
+static uint32_t clks_gating_table[CRU_CLKGATE_CON_CNT] = {
+	/* gate: 0-3 */
+	0xefb8,
+	0x0ff7,
+	0xfff4,
+	0x887f,
+	/* gate: 4-7 */
+	0x0030,
+	0x00f8,
+	0x07e0,
+	0xc000,
+	/* gate: 8-11 */
+	0xff84,
+	0xb047,
+	0x1ca0,
+	0x57ff,
+	/* gate: 12-15 */
+	0x0000,
+	0x00ff,
+	0x1cc0,
+	0x000f,
+};
+
+static void clks_gating(void)
+{
+	uint32_t i;
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(CRU_BASE);
+
+	for (i = 0; i < CRU_CLKGATE_CON_CNT; i++) {
+		dram_d.cru_clkgate[i] = read32(va_base + CRU_CLKGATE_CON(i));
+		write32(BITS_WITH_WMASK(clks_gating_table[i], 0xffff, 0),
+			va_base + CRU_CLKGATE_CON(i));
+	}
+}
+
+static void clks_restore(void)
+{
+	uint32_t i;
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(CRU_BASE);
+
+	for (i = 0; i < CRU_CLKGATE_CON_CNT; i++)
+		write32(BITS_WITH_WMASK(dram_d.cru_clkgate[i], 0xffff, 0),
+			va_base + CRU_CLKGATE_CON(i));
+}
+
+static void pll_power_down(uint32_t pll)
+{
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(CRU_BASE);
+
+	write32(PLL_SLOW_MODE(pll), va_base + CRU_MODE_CON);
+	write32(PLL_POWER_DOWN, va_base + CRU_PLL_CON1(pll));
+}
+
+static void pll_power_up(uint32_t pll)
+{
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(CRU_BASE);
+
+	write32(PLL_POWER_UP, va_base + CRU_PLL_CON1(pll));
+}
+
+static void pll_wait_lock(uint32_t pll)
+{
+	uint32_t loop = 0;
+	vaddr_t va_base = (vaddr_t)phys_to_virt_io(CRU_BASE);
+
+	while (!(read32(va_base + CRU_PLL_CON1(pll)) & PLL_LOCK) &&
+	       (loop < 500)) {
+		udelay(2);
+		loop++;
+	}
+
+	assert(read32(va_base + CRU_PLL_CON1(pll)) & PLL_LOCK);
```

Comment:
Do you mean you prefer panic?
Approved, panic is better, if compiled with NDEBUG not defined, any error generates an assertion failure.